### PR TITLE
Publish courses workflow

### PR DESCRIPTION
**Pull request by the-daniel-rothig**
### Context

Publish workflow:

https://trello.com/c/hzb8hm2o/178-users-submit-courses-for-publication

And preview:

https://trello.com/c/IY8vkQxt/196-course-preview-in-publish-courses
https://trello.com/c/gjumHTBy/151-rendering-markdown-links-lists-and-paragraphs-in-preview

### Changes proposed in this pull request

- add a publish side panel to course variant page that shows publication state and, depending on context, allows you to preview and publish your draft
- Adds a preview screen, taken from Search and Compare, for displaying what the course will look like post-publish

### Guidance to review

This turned out to be a big piece of work, and I expect a number of review comments - guidance for a test strategy and how/where to clarify how the code works would be apprecitated! Still I wanted to raise a PR before the branch goes stale.

Before merging we need to ensure we move the package dependencies from branch-versions to master-versions.

Set the following feature flags

- `features:FEATURE_COURSE_PREVIEW` `True`
- `features:FEATURE_COURSE_PUBLISH` `True`

Of particular note is the CourseMapper, a large but naive piece of domain object mapping that turns a Manage Courses course into a SearchAndCompare course - this might prove useful when publishing courses.

Due to an unfortunate bug in ASP.NET core you will need runtime 2.1.300~2.1.302. The latest (2.1.400) will not work.

### CODE REVIEW

**Comment by the-daniel-rothig** [comment]
[Users submit courses for publication](https://trello.com/c/hzb8hm2o/178-users-submit-courses-for-publication)

**Comment by the-daniel-rothig** [comment]
[Validate 'About your course' mandatory fields on publish](https://trello.com/c/TCcBjbUi/125-validate-about-your-course-mandatory-fields-on-publish)

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,6 @@ node_modules
 *.tmp
 .vs
 dev.targets
```
Comment (defong) [inline .gitignore]: Is that not getting used anymore?

Can you also update readme as well please

Comment (the-daniel-rothig) [reply .gitignore]: done :+1:

File: src/ui/Services/CourseMapper.cs [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                school.Address1,
+                school.Address2,
+                school.Address3,
+                school.Address4
+            }.Where(x => !string.IsNullOrWhiteSpace(x));
+
+            var postCode = school.PostCode ?? "";
+
+            return addressFragments.Any()
+                ? String.Join(", ", addressFragments) + " " + postCode
```
Comment (timabell) [inline src/ui/Services/CourseMapper.cs]: I'm pretty sure you don't need to treat postcode separately, you could just add it to the fragment list.

Minor

File: src/ui/Startup.cs [hunk 3]
```diff
@@ -194,12 +202,14 @@ public void ConfigureServices(IServiceCollection services)
...
 
             services.AddSingleton<IFeatureFlags>(x => new FeatureFlags(Configuration.GetSection("features")));
+            services.AddSingleton<ISearchAndCompareUrlService>(x => new SearchAndCompareUrlService(Configuration.GetValue("SearchAndCompare:UiBaseUrl", "")));
             services.AddSingleton<IManageCoursesApiClientConfiguration, ManageCoursesApiClientConfiguration>();
             services.AddScoped(provider => AnalyticsPolicy.FromEnv());
             services.AddScoped<AnalyticsAttribute>();
             services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
             services.AddSingleton<ManageCoursesConfig, ManageCoursesConfig>();
             services.AddSingleton<IManageApi, ManageApi>();
+            services.AddScoped<ICourseDetailsService>(provider => new CourseDetailsService(provider.GetService<IManageApi>(), new CourseMapper()));
```
Comment (timabell) [inline src/ui/Startup.cs]: I think we can probably get the params of this to be DI'd so you don't need the hard-coded new. Perhaps with the following:

```
services.AddScoped<CourseMapper,CourseMapper>;
services.AddScoped<ICourseDetailsService,CourseDetailsService>;
```

minor

File: src/ui/ViewModels/CourseEnrichmentViewModel.cs [hunk 4]
```diff
@@ -1,14 +1,49 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using GovUk.Education.ManageCourses.Ui.ViewModels.Enums;
 
 namespace GovUk.Education.ManageCourses.Ui.ViewModels
 {
     public class CourseEnrichmentViewModel
```
Comment (timabell) [inline src/ui/ViewModels/CourseEnrichmentViewModel.cs]: some of this file has trailing whitespace, reformat it

minor

File: src/ui/ViewModels/FromUcasViewModel.cs [hunk 5]
```diff
@@ -5,6 +5,9 @@ namespace GovUk.Education.ManageCourses.Ui.ViewModels
 {
     public class FromUcasViewModel
     {
+        public bool AllowPreview { get; set; }
+        public bool AllowPublish { get;  set; }
```
Comment (timabell) [inline src/ui/ViewModels/FromUcasViewModel.cs]: These properties don't quite match up with the class name.

Not sure what to suggest.

Minor.

**Comment by timabell** [comment]
Could this PR have been done in multiple smaller PRs? It's quite hard to review because of the sheer size of the diff. Hindsight etc...

File: src/ui/Services/CourseMapper.cs (hunk 2 above)
Comment (the-daniel-rothig) [reply src/ui/Services/CourseMapper.cs]: no, designs say no comma between address lines and postcodes, hence separate

File: src/ui/ViewModels/FromUcasViewModel.cs (hunk 5 above)
Comment (the-daniel-rothig) [reply src/ui/ViewModels/FromUcasViewModel.cs]: Renamed the class :+1:

File: src/ui/ViewModels/CourseEnrichmentViewModel.cs (hunk 4 above)
Comment (the-daniel-rothig) [reply src/ui/ViewModels/CourseEnrichmentViewModel.cs]: 👍

File: src/ui/Controllers/CourseController.cs [hunk 6]
```diff
@@ -227,7 +274,10 @@ private FromUcasViewModel LoadViewModel(UserOrganisation org, Course course, boo
                 MultipleOrganisations = multipleOrganisations,
                 Course = courseVariant,
                 CourseEnrichment = courseEnrichmentViewModel,
-                RouteData = routeData
+                RouteData = routeData,
+                CurrentUrl = searchAndCompareUrlService.GetCoursePageUri(org.UcasCode, courseVariant.ProgrammeCode),
```
[the remark this answers is not in the document]
Comment (the-daniel-rothig) [reply src/ui/Controllers/CourseController.cs]: Chosen LiveSearchUrl 👍

File: src/ui/Services/CourseMapper.cs (hunk 2 above)
Comment (timabell) [reply src/ui/Services/CourseMapper.cs]: oh it's the comma, missed that. retracted!

**Pull request merged**